batch/test/test_batch.py

```diff
@@ -317,6 +317,173 @@ def assert_batch_ids(expected: Set[int], q=None):
+    b1 = bb1.submit()
+
+    bb2 = create_batch(client, attributes={'tag': tag, 'name': 'b2'})
+    bb2.create_job(DOCKER_ROOT_IMAGE, ['echo', 'test'])
+    b2 = bb2.submit()
+
+    batch_id_test_universe = {b1.id, b2.id}
+
+    def assert_batch_ids(expected: Set[int], q=None):
+        assert expected.issubset(batch_id_test_universe)
+        max_id = max(batch_id_test_universe)
+        min_id = min(batch_id_test_universe)
+        span = max_id - min_id + 1
+        # list_batches returns all batches for all prev run tests so we set a limit
+        batches = client.list_batches(q, last_batch_id=max_id + 1, limit=span, version=2)
+        full_actual = {b.id for b in batches}
+        actual = full_actual.intersection(batch_id_test_universe)
+        assert actual == expected, str((full_actual, max_id, span, b1.debug_info(), b2.debug_info()))
+
+    assert_batch_ids({b1.id, b2.id})
+
+    assert_batch_ids({b1.id, b2.id}, f'tag={tag}')
+    assert_batch_ids({b1.id, b2.id}, f'tag=~{tag}')
+    assert_batch_ids(
+        {b1.id, b2.id},
+        f'''
+name=~b
+tag={tag}
+''',
+    )
+    assert_batch_ids(
+        {b1.id},
+        f'''
+name!~b2
+tag={tag}
+''',
+    )
+    assert_batch_ids(
+        {b1.id},
+        f'''
+name!=b2
+tag={tag}
+''',
+    )
+    assert_batch_ids(
+        {b2.id},
+        f'''
+b2
+tag={tag}
+''',
+    )
+
+    b2.wait()
+
+    assert_batch_ids(
+        {b1.id},
+        f'''
+state != complete
+tag = {tag}
+''',
+    )
+    assert_batch_ids(
+        {b2.id},
+        f'''
+state=complete
+tag={tag}
+''',
+    )
+
+    assert_batch_ids(
+        {b1.id},
+        f'''
+state != success
+tag={tag}
+''',
+    )
+    assert_batch_ids(
+        {b2.id},
+        f'''
+state == success
+tag={tag}
+''',
+    )
+
+    b1.cancel()
+    b1.wait()
+
+    assert_batch_ids(
+        {b1.id},
+        f'''
+state!=success
+tag={tag}
+''',
+    )
+    assert_batch_ids(
+        {b2.id},
+        f'''
+state = success
+tag={tag}
+''',
+    )
+
+    assert_batch_ids(
+        set(),
+        f'''
+state != complete
+tag={tag}
+''',
+    )
+    assert_batch_ids(
+        {b1.id, b2.id},
+        f'''
+state = complete
+tag={tag}
+''',
+    )
+
+    assert_batch_ids(
+        {b2.id},
+        f'''
+tag={tag}
+name=b2
+''',
+    )
+
+    assert_batch_ids(
+        {b2.id},
+        f'''
+tag={tag}
+"b2"
+''',
+    )
+
+    assert_batch_ids(
+        {b2.id},
+        f'''
+tag=~{tag}
+"b2"
+''',
+    )
+
+    assert_batch_ids(
+        {b1.id, b2.id},
+        '''
+start_time >= 2023-02-24T17:15:25Z
+end_time < 3000-02-24T17:15:25Z
+''',
+    )
+
+    assert_batch_ids(
+        set(),
+        '''
+start_time >= 2023-02-24T17:15:25Z
+end_time == 2023-02-24T17:15:25Z
+''',
+    )
+
+    assert_batch_ids(set(), 'duration > 50000')
+    assert_batch_ids(set(), 'cost > 1000')
+    assert_batch_ids({b1.id}, f'batch_id = {b1.id}')
+    assert_batch_ids({b1.id}, f'batch_id == {b1.id}')
+
```

Comment:
I think these tests are good, but it's hard to make them really thorough. Maybe not in this PR but as part of this project we can write some tests that just test the parser, and run it through some exhaustive tests to make sure that every possible grammatical construct parses without error (or with error for queries that don't support certain operators). Such tests would be super cheap and give me more confidence that all of the exception handling is working properly.